check_py_coverage.sh

```diff
@@ -0,0 +1,74 @@
+#!/bin/bash
+SCRIPT_DIRNAME=`dirname $0`
+SCRIPT_DIR=`(cd $SCRIPT_DIRNAME; pwd)`
+source $SCRIPT_DIR/raven/scripts/establish_conda_env.sh --quiet --load
+RAVEN_LIBS_PATH=`conda env list | awk -v rln="$RAVEN_LIBS_NAME" '$0 ~ rln {print $NF}'`
+BUILD_DIR=${BUILD_DIR:=$RAVEN_LIBS_PATH/build}
+INSTALL_DIR=${INSTALL_DIR:=$RAVEN_LIBS_PATH}
+PYTHON_CMD=${PYTHON_CMD:=python}
+JOBS=${JOBS:=1}
+mkdir -p $BUILD_DIR
+mkdir -p $INSTALL_DIR
+DOWNLOADER='curl -C - -L -O '
+
+ORIGPYTHONPATH="$PYTHONPATH"
+
+update_python_path ()
+{
+    if ls -d $INSTALL_DIR/lib/python*
+    then
+        export PYTHONPATH=`ls -d $INSTALL_DIR/lib/python*/site-packages/`:"$ORIGPYTHONPATH"
+    fi
+}
+
+update_python_path
+PATH=$INSTALL_DIR/bin:$PATH
+
+if which coverage
+then
+    echo coverage already available, skipping building it.
+else
+    if curl http://www.energy.gov > /dev/null
+    then
+       echo Successfully got data from the internet
+    else
+       echo Could not connect to internet
+    fi
+
+    cd $BUILD_DIR
+    #SHA256=56e448f051a201c5ebbaa86a5efd0ca90d327204d8b059ab25ad0f35fbfd79f1
+    $DOWNLOADER https://files.pythonhosted.org/packages/ef/05/31553dc038667012853d0a248b57987d8d70b2d67ea885605f87bcb1baba/coverage-7.5.4.tar.gz
+    tar -xvzf coverage-7.5.4.tar.gz
+    cd coverage-7.5.4
+    (unset CC CXX; $PYTHON_CMD setup.py install --prefix=$INSTALL_DIR)
+fi
+
+update_python_path
+
+cd $SCRIPT_DIR
+
+#coverage help run
+SRC_DIR=`(cd src && pwd)`
+
+# get display var
+DISPLAY_VAR=`(echo $DISPLAY)`
+# reset it
+export DISPLAY=
+
+EXTRA="--rcfile=$SRC_DIR/../tests/.coveragerc --source=$SRC_DIR --parallel-mode"
+export COVERAGE_FILE=`pwd`/.coverage
+
+coverage erase --rcfile="$SRC_DIR/../tests/.coveragerc"
+($SRC_DIR/../run_tests "$@" --python-command="coverage run $EXTRA " || echo run_test done but some tests failed)
+
+#get DISPLAY BACK
+DISPLAY=$DISPLAY_VAR
+
+## Go to the final directory and generate the html documents
+cd $SCRIPT_DIR/tests/
+pwd
+rm -f .cov_dirs
+for FILE in `find . -name '.coverage.*'`; do dirname $FILE; done | sort | uniq > .cov_dirs
```

Comment:
it looks like we move to the `/FORCE/tests` directory in Line 68 but then we try to call the `.coverage` file which should be in the parent `/FORCE` directory. would this cause a problem?

Reply:
After studying these lines, you are correct that they do not make much sense. Strangely, they do not cause a problem given the current setup, however. The for loop searches for coverage information files within the tests directory. It finds none because the .coverage.* files are all directly under FORCE. Thus .cov_dirs is still empty after line 71. In line 72, when `cat .cov_dirs` is called, it returns nothing (or whitespace; it doesn't really matter which). Thus the command remaining in 72 is just `coverage combine`. This happens to be a perfectly valid command for coverage, and is designed as a higher level combination command that tries to find a single directory where all the .coverage.* files have been manually collected by the user. While there is no online documentation speaking to this theory, it would seem that, based on the behavior of the script, one of the places searched by `coverage combine` is the directory of the shell script in which it is being run. 

While the current code is functional, it should certainly be changed. Preserving the flexibility of the script in identifying .coverage.* files in a variety of locations would be convenient; however, it could also introduce unnecessary risk. As long as all the tests are run via `FORCE/run_tests`, all the tests will be in a single directory, so the simple `coverage combine` command should work fine. If the alternative solution is pursued, combine has a potential to find .coverage.* files in other locations that were not produced during the most recent run. Typically, the `coverage erase` command in line 61 should take care of this, but it is doubtful whether its flexibility is sufficient to protect the same region that the combine command may search. In addition, protecting such a large region could cause unwanted deletion of other, unrelated .coverage.* files that happen to be in the line of fire. Thus, I believe the safest solution is to remain in the FORCE directory and use the high-level `coverage combine` command.

Reply:
ok, just making sure that there wasn't an obvious error here. I think we can leave it as is, as you've recommended here.